zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1073,10 +1073,24 @@ protected void processStateChanged(WatchedEvent event) {
 
     fireStateChangedEvent(event.getState());
 
+    // This following is the case for sharedZkClient
     if (!isManagingZkConnection()) {
+      /*
+       For SharedZkClient, we will not invoke fireNewSessionEvents and we will invoke fireAllEvents
+       This keeps the same behavior as original SharedZkClient.
+       Invoking fireAllEvents is important to make sure HelixPropertyStore working correctly.
+      */
+      if ((event.getState() == KeeperState.SyncConnected) && (!_isNewSessionEventFired) && (!"0"
+          .equals(getHexSessionId()))) {
+        fireAllEvents();
+        _isNewSessionEventFired = true;
+      } else if (event.getState() == KeeperState.Expired) {
+        _isNewSessionEventFired = false;
+      }
```

Comment:
If we remove ` _isNewSessionEventFired = false;` in `reconnect()` and do below code, it seems more concise and easier to maintain. `_isNewSessionEventFired` is only changed in this code block. What do you think?

```java
    if (event.getState() == KeeperState.SyncConnected && !_isNewSessionEventFired && !"0".equals(getHexSessionId())) {
      if (isManagingZkConnection()) {
        /*
         * Before the new zookeeper instance is connected to the zookeeper service and its session
         * is established, its session id is 0.
         * New session event is not fired until the new zookeeper session receives the first
         * SyncConnected state(the zookeeper session is established).
         * Now the session id is available and non-zero, and we can fire new session events.
         */
        fireNewSessionEvents();
      }
      /*
       * Set it true to avoid firing events again for the same session next time
       * when SyncConnected events are received.
       */
      _isNewSessionEventFired = true;

      /*
       * With this first SyncConnected state, we just get connected to zookeeper service after
       * reconnecting when the session expired. Because previous session expired, we also have to
       * notify all listeners that something might have changed.
       */
      fireAllEvents();
    } else if (event.getState() == KeeperState.Expired) {
      if (isManagingZkConnection()) {
        reconnectOnExpiring();
      }
       _isNewSessionEventFired = false;
    }
```

Reply:
Good point. 
Yes, this is another way to do it, thought about it.
Pro: concise
Con: hard to understand the intention. This may pave the way to another hidden bug when people do refactoring next time, as how we get to this issue.

Reply:
To me, it is easier to understand:
1. only managed connection will do: fireNewSessionEvents() and reconnectOnExpiring.
2. first sync connected: _isNewSessionEventFired = true; expired: _isNewSessionEventFired = false.

Reply:
@jiajunwang, what is your take here?

Reply:
I like less code : )

And actually I can understand the concise version. So I guess it's better.

Reply:
The I will use this concise version. I will add a verbose note about the intention too.

Reply:
By they way, setting _isNewSessionEventFired = false should be above reconnect(). Otherwise, there is a small chance of race condition. In reconnect() or current code, we still have this race condition.
```
    } else if (event.getState() == KeeperState.Expired) {
       _isNewSessionEventFired = false;
      if (isManagingZkConnection()) {
        reconnectOnExpiring();
      }
    }
```